lib/puppet/provider/package/version/range.rb

```diff
@@ -0,0 +1,102 @@
+require 'puppet/provider/package/version'
+class Puppet::Provider::Package::Version::Range
+  class ValidationFailure < ArgumentError; end
+  # Parses a version range string into a comparable {Range} instance.
+  #
+  # Currently parsed version range string may take any of the following:
+  # forms:
+  #
+  # * Regular Version strings
+  #   * ex. `"1.0.0"`, `"1.2.3-pre"`
+  # * Inequalities
+  #   * ex. `">1.0.0"`, `"<3.2.0"`, `">=4.0.0"`
+  # * Range Intersections (min is always first)
+  #   * ex. `">1.0.0 <=2.3.0"`
+  #
+  RANGE_SPLIT = /\s+/
+  FULL_REGEX = /\A((?:[<>=])+)(.+)\Z/
+
+  # @param range_string [String] the version range string to parse
+  # @return [Range] a new {Range} instance
+  # @api public
+  def self.parse(range_string, version_class)
+    simples = range_string.split(RANGE_SPLIT).map do |simple|
+      match, operator, version = *simple.match(FULL_REGEX)
+      raise ValidationFailure, "Unable to parse '#{simple}' as a version range identifier" unless match
+      case operator
+      when '>'
+        GtRange.new(version_class::parse(version))
+      when '>='
+        GtEqRange.new(version_class::parse(version))
+      when '<'
+        LtRange.new(version_class::parse(version))
+      when '<='
+        LtEqRange.new(version_class::parse(version))
+      else
+        raise ValidationFailure, "Operator '#{operator}' is not implemented"
+      end
+    end
+    simples.size == 1 ? simples[0] : MinMaxRange.new(simples[0], simples[1])
+  end
+
+  class MinMaxRange
```

Comment:
can we move this small classes to their specific files?
```
lib/puppet/provider/package/version/ranges/min_max_range.rb```

Reply:
I suppose this means tests will also have to be split

Reply:
We can still test `Range` as a whole(as it is now), since there is not much logic in those small classes.